### PR TITLE
Basic BLS-12-381-pk-in-g1 verify

### PR DESCRIPTION
**Pull request by benr-ml**

### CODE REVIEW

File: fastcrypto/src/bls12381.rs [hunk 1]
```diff
@@ -48,7 +49,7 @@ pub const DST: &[u8] = b"BLS_SIG_BLS12381G1_XMD:SHA-256_SSWU_RO_NUL_";
 #[readonly::make]
 #[derive(Default, Clone)]
 pub struct BLS12381PublicKey {
-    pub pubkey: blst::PublicKey,
+    pub pubkey: blst_min_sig::PublicKey,
```
[the remark this answers is not in the document]
Comment (benr-ml) [reply fastcrypto/src/bls12381.rs:52]: that's the same object, I didn't change the type just renamed the module

File: fastcrypto/src/bls12381.rs [hunk 2]
```diff
@@ -693,10 +696,24 @@ impl Drop for BLS12381KeyPair {
...
+// Low level BLS verification for when the public key is in G_1.
+pub fn bls12_381_pk_in_g1_verify(
+    msg: &[u8],
+    pk: &[u8],
+    sig: &[u8],
+) -> Result<bool, FastCryptoError> {
+    const G2_DST: &[u8] = b"BLS_SIG_BLS12381G2_XMD:SHA-256_SSWU_RO_NUL_";
+    let sig = blst_min_pk::Signature::from_bytes(sig).map_err(|_| FastCryptoError::InvalidInput)?;
+    let pk = blst_min_pk::PublicKey::from_bytes(pk).map_err(|_| FastCryptoError::InvalidInput)?;
+    let err = sig.verify(true, msg, G2_DST, &[], &pk, true);
```
Comment (kchalkias) [inline fastcrypto/src/bls12381.rs:717]: So interesting that verify returns an Error object, even on success.

**Review by kchalkias — COMMENTED** [review]
I think we should have two different BLS sig scheme structs for the two G1/G2 options for readability and avoid any misuse.

**Comment by benr-ml** [comment]
> I think we should have two different BLS sig scheme structs for the two G1/G2 options for readability and avoid any misuse.

@kchalkias Do you mean to duplicate all the structs we have there now? The reason I chose not to do that was that we do not plan to use this variant of BLS anywhere besides Move, so I'd prefer not to maintain lots of code unless needed. Once we need this variant in our protocols, we could obviously add it and switch.

**Review by kchalkias — COMMENTED** [review]
@benr-ml The reality is eventually we'd need a rich api that verifies both modes. For instance even for logging purposes we want to have a Rust api to debug etc. That said IMO these are different algorithms from the dev's point of view. It won't only be Drand and co, there exist partners that use G1 for keys as well and it might be required for bridges.
I think it's more clear to handle it as a different algorithm.

**Review by huitseeker — CHANGES_REQUESTED** [review]
This is switching the behavior of the existing BLS12-381 used in Sui, Narwhal, etc.
I would recommend implementing a distinct scheme (in the sense of a distinct Rust module) instead of modifying the behavior of this library for downstream dependents.

**Comment by benr-ml** [comment]
@huitseeker - it doesn't change the current behavior, it just renames the module name used by the existing code. 

@kchalkias - sure, I'll add the other structs tomorrow.

**Comment by huitseeker** [comment]
@benr-ml oh, yes! 🤦 sorry, I read too fast and thought you were testing `min_pk` here.

**Pull request closed without merging**